### PR TITLE
Added rest layer for the workflow. Added secure tests

### PR DESCRIPTION
**Pull request by stevanbz**
*Issue #, if available:*

*Description of changes:*
Added REST API layer for workflow:
- API for create/update workflow - POST/PUT /_plugins/_alerting/workflows
- API for getting the workflow - GET /_plugins/_alerting/workflows/{workflowID}
- API for deleting the workflow - DELETE /_plugins/_alerting/workflows/{workflowID}

Beside REST API, secure tests are added - in order to test the workflow and underlying monitors (delegates):
```shell
 ./gradlew integTestRemote -xktlint -Dtest.debug=1  -Dcluster.debug=1 --tests "org.opensearch.alerting.resthandler.SecureWorkflowRestApiIT" -Dtests.rest.cluster=localhost:9200 -Dtests.cluster=localhost:9200 -Dtests.clustername=docker-cluster -Dsecurity=true -Dhttps=true -Duser=admin -Dpassword=admin
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 7.4.2
  OS Info               : Linux 5.15.0-67-generic (amd64)
  JDK Version           : 11 (Eclipse Temurin JDK)
  JAVA_HOME             : /home/stevan/.asdf/installs/java/adoptopenjdk-11.0.15+10
  Random Testing Seed   : 25C8D7B2BE3426BC
  In FIPS 140 mode      : false
=======================================

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.4.2/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 2m 2s
12 actionable tasks: 1 executed, 11 up-to-date
``` 
List of the tests:

- test create workflow with an user with alerting role - creates user with hr backend role and index workflow access; verifies that user can create a workflow
- test create workflow with an user without alerting role - creates user with hr backend role and without index workflow access; verifies that 403 status is returned when creating a workflow
- test create workflow with an user with read-only role 
- test create workflow with delegate with an user without index read role - creates user with index workflow cluster permission and with hr_data index access; creates a monitor for not_hr_data index; tries creating a workflow and verifies that user can't access the monitor indices
- test create workflow with disable filter by
- test get workflow with an user with get workflow role - tries to get the workflow with the user with get workflow cluster privilege
- test get workflow with an user without get monitor role - verifies that user without get workflow cluster privilege can't access the created workflow
- test update workflow with disable filter by - tries to update a workflow as admin with disabled filter by
- test update workflow with enable filter by - tries to update a workflow as admin with enabled filter by
- test create workflow with **enable filter by** with a user have access and without role has no access - creates a workflow with user having two backend roles (role2, HR) with alerting_full_access cluster privileges; verifies that user which has two backend roles (role2, HR) get workflow access cluster privilege; updates the user roles to role1 and verifies that user can't get the workflow
- test create workflow with **enable filter by** with a user with a backend role doesn't have access to monitor - monitor is created by the user which has two backend roles - hr and role2; tries to create a workflow for the previously created monitor as delegate with a user with different backend roles (role3);
- test create workflow with enable filter by with no backend roles
- test create workflow as admin with enable filter by with no backend roles - creates workflow as admin and verifies that user with specified backend roles can't get the workflow
- test create workflow with enable filter by with roles user has no access and throw exception 
- test create workflow as admin with enable filter by with a user have access and without role has no access - creates a workflow with 3 backend roles; checks if the user with one backend role matching the previous three can access the monitor; updates the user by setting up a new backend role - role5 and verifies that he can't access the monitor
- test update workflow with enable filter by with removing a permission - creates a workflow with one set of backend roles; updates the workflow by removing one of the backend roles; verifies that user can't access (get) the workflow
- test update workflow with enable filter by with no backend roles
- test update workflow as admin with enable filter by with no backend roles - creates workflow with specified backend roles and verifies that user with matching roles can access the workflow; updates the workflow backend role can access the workflow (GET); 
removes the backend role and verifies that user can't access  anymore to a workflow
- test update workflow with enable filter by with updating with a permission user has no access to and throw exception - verifies that updating a workflow with a new backend role, with a user which doesn't have the same backend role assigned, will fail
- test update workflow as another user with enable filter by with removing a permission and adding permission - same as above, just user has the role on which he updates the workflow
- test update workflow as admin with enable filter by with removing a permission
- test delete workflow with **disable filter by** - verifies deleting the workflow with it's delegates
- test delete workflow with **enable filter by**
- test delete workflow with **enable filter** with user that doesn't have delete_monitor cluster privilege failed 
- test execute workflow with an user with execute workflow access 
- test execute workflow with an user without execute workflow access
- test delete workflow with an user with delete workflow access 
- test delete workflow with deleting delegates with an user with delete workflow access
- test delete workflow with an user without delete monitor access
- test admin all access with enable filter by - verifies that admin can create workflow, get it and remove it when filter by is enabled
- test execute workflow with bucket-level and doc-level chained monitors with user having partial index permissions

*CheckList:*
[ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by lezzago** [comment]
Please provide proof of security tests passing

File: alerting/src/main/kotlin/org/opensearch/alerting/resthandler/RestGetWorkflowAction.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            RestHandler.Route(
+                RestRequest.Method.GET,
+                "${AlertingPlugin.WORKFLOW_BASE_URI}/{workflowID}"
+            ),
+            RestHandler.Route(
+                RestRequest.Method.HEAD,
+                "${AlertingPlugin.WORKFLOW_BASE_URI}/{workflowID}"
+            )
+        )
+    }
```
[the remark this answers is not in the document]
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/resthandler/RestGetWorkflowAction.kt:42]: For checking if the workflow exists

File: alerting/src/main/kotlin/org/opensearch/alerting/resthandler/RestGetWorkflowAction.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    override fun prepareRequest(request: RestRequest, client: NodeClient): RestChannelConsumer {
+        log.debug("${request.method()} ${AlertingPlugin.WORKFLOW_BASE_URI}/{workflowID}")
+
+        val workflowId = request.param("workflowID")
+        if (workflowId == null || workflowId.isEmpty()) {
+            throw IllegalArgumentException("missing id")
+        }
+
+        var srcContext = context(request)
+        if (request.method() == RestRequest.Method.HEAD) {
```
Comment (lezzago) [inline alerting/src/main/kotlin/org/opensearch/alerting/resthandler/RestGetWorkflowAction.kt:49]: We need to add the route for HEAD. Sorry about making it get removed. Lets add the HEAD route.

File: alerting/src/test/kotlin/org/opensearch/alerting/AccessRoles.kt [hunk 3]
```diff
@@ -30,9 +37,13 @@ val ROLE_TO_PERMISSION_MAPPING = mapOf(
     ALERTING_SEARCH_EMAIL_GROUP_ACCESS to "cluster:admin/opendistro/alerting/destination/email_group/search",
     ALERTING_INDEX_MONITOR_ACCESS to "cluster:admin/opendistro/alerting/monitor/write",
     ALERTING_GET_MONITOR_ACCESS to "cluster:admin/opendistro/alerting/monitor/get",
+    ALERTING_GET_WORKFLOW_ACCESS to AlertingActions.GET_WORKFLOW_ACTION_NAME,
```
Comment (lezzago) [inline alerting/src/test/kotlin/org/opensearch/alerting/AccessRoles.kt:40]: As a followup, we should also update the other cluster permissions to point to the action names as well.

File: alerting/src/test/kotlin/org/opensearch/alerting/resthandler/SecureWorkflowRestApiIT.kt [hunk 4]
```diff
@@ -0,0 +1,1406 @@
...
+
+        try {
+            val deleteWorkflowResponse = userClient?.makeRequest(
+                "DELETE",
+                "$WORKFLOW_ALERTING_BASE_URI/${workflow.id}?refresh=$refresh",
+                emptyMap(),
+                monitor.toHttpEntity()
+            )
+            assertEquals("DELETE workflow failed", RestStatus.OK, deleteWorkflowResponse?.restStatus())
+        } finally {
```
Comment (lezzago) [inline alerting/src/test/kotlin/org/opensearch/alerting/resthandler/SecureWorkflowRestApiIT.kt:1168]: If we are doing the try and finally, we need to have a catch for any exception and fail on that. Or we dont do the try finally part. This is same with the other tests like this.

**Review by lezzago — APPROVED** [review]
LGTM with some minor issues that could be followed up in another PR

**Review by eirsep — APPROVED** [review]

**Pull request merged**